File: week4/python/ReverseNodesInK_Group_25.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        if not head:
            return head
        
        node = head
        stack = []
        for _ in range(k):
            if not node:
                break
                
            stack.append(node)
            node = node.next
        
        if len(stack) == k:
            head_group = stack.pop()
            cur = head_group

            while stack:
                cur.next = stack.pop()
                cur = cur.next
            
            cur.next = self.reverseKGroup(node, k)
            
        else:
            head_group = stack[0] if stack else None
            
        return head_group
                

File: week4/java/ReverseNodesInK_Group_25.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode reverseKGroup(ListNode head, int k) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode cur = head;
        int count = 0;
        while (cur != null && count < k) {
            count += 1;
            cur = cur.next;
        }
        if (count == k) {
            ListNode prev = null;
            ListNode lastNode = head;
            while (count > 0) {
                ListNode tmp = head.next;
                head.next = prev;
                prev = head;
                head = tmp;
                count -= 1;
            }
            lastNode.next = reverseKGroup(cur, k);
            return prev;
        }
        return head;
    }
}
